flink-filesystems/flink-s3-fs-base/src/main/java/org/apache/flink/fs/s3/common/FlinkS3FileSystem.java

```diff
@@ -101,13 +237,154 @@ public FlinkS3FileSystem(
         this.maxConcurrentUploadsPerStream = maxConcurrentUploadsPerStream;
+        LOG.info("Created Flink S3 FS, s5Cmd configuration: {}", s5CmdConfiguration);
     }
 
     // ------------------------------------------------------------------------
 
+    @Override
+    public boolean canCopyPaths(Path source, Path destination) {
+        return canCopyPaths();
+    }
+
+    private boolean canCopyPaths() {
+        return s5CmdConfiguration != null;
+    }
+
+    @Override
+    public void copyFiles(List<CopyRequest> requests, ICloseableRegistry closeableRegistry)
+            throws IOException {
+        checkState(canCopyPaths(), "#downloadFiles has been called illegally");
+        List<String> artefacts = new ArrayList<>();
+        artefacts.add(s5CmdConfiguration.path);
+        artefacts.addAll(s5CmdConfiguration.args);
+        artefacts.add("run");
+        castSpell(convertToSpells(requests), closeableRegistry, artefacts.toArray(new String[0]));
+    }
+
+    private List<String> convertToSpells(List<CopyRequest> requests) throws IOException {
+        List<String> spells = new ArrayList<>();
+        for (CopyRequest request : requests) {
+            Files.createDirectories(Paths.get(request.getDestination().toUri()).getParent());
+            spells.add(
+                    String.format(
+                            "cp %s %s",
+                            request.getSource().toUri().toString(),
+                            request.getDestination().getPath()));
+        }
+        return spells;
+    }
+
+    private void castSpell(
+            List<String> spells, ICloseableRegistry closeableRegistry, String... artefacts)
+            throws IOException {
+        LOG.info("Casting spell: {}", Arrays.toString(artefacts));
+        int exitCode = 0;
+        final AtomicReference<IOException> maybeCloseableRegistryException =
+                new AtomicReference<>();
+
+        // Setup temporary working directory for the process
+        File tmpWorkingDir = new File(localTmpDir, "s5cmd_" + UUID.randomUUID());
+        java.nio.file.Path tmpWorkingPath = Files.createDirectories(tmpWorkingDir.toPath());
+
+        try {
+            // Redirect the process input/output to files. Communicating directly through a
+            // stream can lead to blocking and undefined behavior if the underlying process is
+            // killed (known Java problem).
+            ProcessBuilder hogwart = new ProcessBuilder(artefacts).directory(tmpWorkingDir);
+            s5CmdConfiguration.configureEnvironment(hogwart.environment());
+            File inScrolls = new File(tmpWorkingDir, "s5cmd_input");
+            Preconditions.checkState(inScrolls.createNewFile());
+            File outScrolls = new File(tmpWorkingDir, "s5cmd_output");
+            Preconditions.checkState(outScrolls.createNewFile());
+
+            FileUtils.writeFileUtf8(inScrolls, String.join(System.lineSeparator(), spells));
```

Comment:
I think,
a line separator after the last string is necessary
because the file content serves as input to a process
and similar to input from a terminal it needs a newline to take
effect